cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -0,0 +1,231 @@
+* Reusing the same stream will be effiecient in terms of resources as we are not creating/destroying the stream for every update.
+
+##### Cons
+
+* An overhead on the cilium agent to keep track of the streams of the connected SDP instances.
+* Streams are not thread safe, so if we have multiple threads using the same stream we will need to handle the synchronization.
+* If SDP in future decides to handle policy updates as well, it will be tricky. SDP needs to keep trying to sync with CA as well as CA needs to keep a track of the SDP instances to send updates.
+
+![standalone-dns-proxy-up-scenario](./images/standalone-dns-proxy-up-scenario.png)
+![standalone-dns-proxy-down-scenario](./images/standalone-dns-proxy-down-scenario.png)
+
+### Reading from file system on startup
+
+SDP can read from the file system(`ep_config.json`) and get the DNS Rules on bootup. In case, it is able to connect to cilium agent, cilium agent will send the current snapshot of the dns rules to sdp. In case, cilium agent is down, SDP can continue serving the DNS requests based on DNS rules retrieved from the filesystem on bootup.
+
+### Q : Reason behind lazily updating the DNS rules in the ep_config.json file?
+
+### Discovering endpoint metadata from SDP
+
+In order to enforce L7 DNS policy, SDP needs to translate IP address to endpoint ID and identity ID. The simplest option is to reuse the same gRPC stream and implement another method to get the mappings from cilium agent. However, the actual source of truth are the bpf maps and SDP is expected to resolve the DNS queries even when agent is down, so we could read the information directly from the bpf maps. We prefer the bpf option since it does not rely on agent availability.
+
+#### Option 1a: gRPC method
+
+Get the endpoint ID and identity for a given IP address by making a gRPC call to cilium agent.
+
+##### Pros
+
+* Avoid interacting with low-level details. Simpler to maintain and stay upto date with any datapath changes.
+* All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
+
+##### Cons
+
+* Reliance on gRPC call in the hot-path
+* Does not cache the data for endpoints that never made a DNS query.
+* In an event where SDP restarts when agent's gRPC service is unavailable, all cached state is lost and SDP cannot translate IP to endpoint ID or identity.
+
+#### Option 1b: Listen to ipcache updates via grpc
+
+Similar to envoy, SDP can listen to ipcache updates and maintain a local cache of IP to endpoint ID mappings via grpc. This changes the way DNS proxy currently behaves i.e pull based model to push based model.
+
+##### Pros
+
+* Simpler to maintain and stay upto date with any new ip<>identity changes.
+* All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
+
+##### Cons
+
+* Need a mechanism for reconciling the cache in case of SDP restarts.
```

Comment:
Yes, con of this specific design. The solution is to send the current state to the listener when it starts listening. But since we recommend reading from bpf maps, this won't be needed.
Also we kept the options we considered for getting the data(endpoint id and identity). Let us know if we need to remove those options and focus on what we want to propose.
SDP will be reading from the state files for reconciliation at the start.